File: app/src/main/java/app/trigger/HttpsDoorSetup.java
package app.trigger;

import android.graphics.Bitmap;
import android.graphics.drawable.Icon;

import app.trigger.DoorState.StateCode;

import java.security.cert.Certificate;


public class HttpsDoorSetup implements Setup {
    static final String type = "HttpsDoorSetup";
    int id;
    String name;
    public Boolean require_wifi;
    public String method;
    public String open_query;
    public String close_query;
    public String ring_query;
    public String status_query;

    // regex to evalute the door return message
    public String unlocked_pattern;
    public String locked_pattern;

    public Bitmap open_image;
    public Bitmap closed_image;
    public Bitmap unknown_image;
    public Bitmap disabled_image;

    public String ssids;
    public Certificate certificate;
    public Boolean ignore_certificate;
    public Boolean ignore_hostname_mismatch;
    public Boolean ignore_expiration;

    public HttpsDoorSetup(int id, String name) {
        this.id = id;
        this.name = name;
        this.require_wifi = true;
        this.method = "";
        this.open_query = "";
        this.close_query = "";
        this.ring_query = "";
        this.status_query = "";
        this.locked_pattern = "LOCKED";
        this.unlocked_pattern = "UNLOCKED";
        this.open_image = null;
        this.closed_image = null;
        this.unknown_image = null;
        this.disabled_image = null;
        this.ssids = "";
        this.certificate = null;
        this.ignore_certificate = false;
        this.ignore_hostname_mismatch = false;
        this.ignore_expiration = false;
    }

    @Override
    public int getId() {
        return id;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String getType() {
        return type;
    }

    @Override
    public String getSSIDs() {
        return ssids;
    }

    @Override
    public Bitmap getStateImage(StateCode state) {
        switch (state) {
            case OPEN:
                return open_image;
            case CLOSED:
                return closed_image;
            case DISABLED:
                return disabled_image;
            case UNKNOWN:
                return unknown_image;
            default:
                return null;
        }
    }

    private static String stripUrls(String... urls) {
        // remove path
        String prefix = "https://";
        for (String url : urls) {
            if (url.startsWith(prefix)) {
                int i = url.indexOf('/', prefix.length());
                if (i > 0) {
                    url = url.substring(0, i);
                }
            }
            return url;
        }

        return "";
    }

    @Override
    public String getRegisterUrl() {
        // extract from known urls
        return stripUrls(open_query, ring_query, close_query, status_query);
    }

    @Override
    public DoorState parseReply(DoorReply reply) {
        return Utils.genericDoorReplyParser(reply, this.unlocked_pattern, this.locked_pattern);
    }

    @Override
    public boolean canOpen() {
        return !Utils.isEmpty(open_query);
    }

    @Override
    public boolean canClose() {
        return !Utils.isEmpty(close_query);
    }

    @Override
    public boolean canRing() {
        return !Utils.isEmpty(ring_query);
    }
}

File: app/src/main/kotlin/app/trigger/HttpsDoorSetup.kt
package app.trigger

import android.graphics.Bitmap
import app.trigger.DoorState.StateCode
import app.trigger.https.HttpsTools
import app.trigger.ssh.KeyPairBean
import app.trigger.ssh.SshTools
import java.security.cert.Certificate
import org.json.JSONObject


class HttpsDoorSetup(override var id: Int, override var name: String) : Setup {
    override val type = Companion.type

    var require_wifi = false
    var method = "GET"
    var open_query = ""
    var close_query = ""
    var ring_query = ""
    var status_query = ""
    var ssids = ""

    // regex to evaluate the door return message
    var unlocked_pattern = "UNLOCKED"
    var locked_pattern = "LOCKED"
    var open_image: Bitmap? = null
    var closed_image: Bitmap? = null
    var unknown_image: Bitmap? = null
    var disabled_image: Bitmap? = null

    var server_certificate: Certificate? = null
    var client_certificate: Certificate? = null
    var client_keypair: KeyPairBean? = null
    var ignore_certificate = false
    var ignore_hostname_mismatch = false
    var ignore_expiration = false

    override fun getWiFiRequired(): Boolean = require_wifi
    override fun getWiFiSSIDs(): String = ssids

    override fun getStateImage(state: StateCode?): Bitmap? {
        return when (state) {
            StateCode.OPEN -> open_image
            StateCode.CLOSED -> closed_image
            StateCode.DISABLED -> disabled_image
            StateCode.UNKNOWN -> unknown_image
            else -> null
        }
    }

    // extract from known urls
    override fun getRegisterUrl(): String {
        return stripUrls(open_query, ring_query, close_query, status_query)
    }

    override fun parseReply(reply: DoorReply): DoorState {
        return Utils.genericDoorReplyParser(reply, unlocked_pattern, locked_pattern)
    }

    override fun canOpen(): Boolean {
        return !Utils.isEmpty(open_query)
    }

    override fun canClose(): Boolean {
        return !Utils.isEmpty(close_query)
    }

    override fun canRing(): Boolean {
        return !Utils.isEmpty(ring_query)
    }

    fun toJSONObject(): JSONObject {
        val obj = JSONObject()
        obj.put("id", id)
        obj.put("name", name)
        obj.put("type", type)
        obj.put("require_wifi", require_wifi)
        obj.put("method", method)
        obj.put("open_query", open_query)
        obj.put("close_query", close_query)
        obj.put("ring_query", ring_query)
        obj.put("status_query", status_query)
        obj.put("ssids", ssids)
        obj.put("unlocked_pattern", unlocked_pattern)
        obj.put("locked_pattern", locked_pattern)

        obj.put("open_image", Utils.serializeBitmap(open_image))
        obj.put("closed_image", Utils.serializeBitmap(closed_image))
        obj.put("unknown_image", Utils.serializeBitmap(unknown_image))
        obj.put("disabled_image", Utils.serializeBitmap(disabled_image))

        obj.put("server_certificate", HttpsTools.serializeCertificate(server_certificate))
        obj.put("client_certificate", HttpsTools.serializeCertificate(client_certificate))
        obj.put("client_keypair", SshTools.serializeKeyPair(client_keypair))

        obj.put("ignore_certificate", ignore_certificate)
        obj.put("ignore_hostname_mismatch", ignore_hostname_mismatch)
        obj.put("ignore_expiration", ignore_expiration)

        return obj
    }

    companion object {
        const val type = "HttpsDoorSetup"

        fun fromJSONObject(obj: JSONObject): HttpsDoorSetup {
            val id = obj.getInt("id")
            val name = obj.getString("name")
            val setup = HttpsDoorSetup(id, name)

            setup.require_wifi = obj.optBoolean("require_wifi", false)
            setup.method = obj.optString("method", "GET")
            setup.open_query = obj.optString("open_query", "")
            setup.close_query = obj.optString("close_query", "")
            setup.ring_query = obj.optString("ring_query", "")
            setup.status_query = obj.optString("status_query", "")
            setup.ssids = obj.optString("ssids", "")
            setup.unlocked_pattern = obj.optString("unlocked_pattern", "UNLOCKED")
            setup.locked_pattern = obj.optString("locked_pattern", "LOCKED")

            setup.open_image = Utils.deserializeBitmap(obj.optString("open_image", ""))
            setup.closed_image = Utils.deserializeBitmap(obj.optString("closed_image", ""))
            setup.unknown_image = Utils.deserializeBitmap(obj.optString("unknown_image", ""))
            setup.disabled_image = Utils.deserializeBitmap(obj.optString("disabled_image", ""))
            setup.server_certificate = HttpsTools.deserializeCertificate(obj.optString("server_certificate", ""))
            setup.client_certificate = HttpsTools.deserializeCertificate(obj.optString("client_certificate", ""))
            setup.client_keypair = SshTools.deserializeKeyPair(obj.optString("client_keypair", ""))

            setup.ignore_certificate = obj.optBoolean("ignore_certificate", false)
            setup.ignore_hostname_mismatch = obj.optBoolean("ignore_hostname_mismatch", false)
            setup.ignore_expiration = obj.optBoolean("ignore_expiration", false)

            return setup
        }

        private fun stripUrls(vararg urls: String): String {
            // remove path
            val prefix = "https://"
            for (url in urls) {
                if (url.startsWith(prefix)) {
                    val i = url.indexOf('/', prefix.length)
                    if (i > 0) {
                        return url.substring(0, i)
                    }
                }
                return url
            }
            return ""
        }
    }
}
